xStation initial support

bbsan2k opened the pull request:
This PR adds initial support for xstation and Game ID parsing.
Additional stuff that needs to be implemented is a game database containing multi disc matching and game names for proper displaying.

xyzz reviewed (changes requested):
Okay, here are some initial comments from reading the code:

- there are too many new buffers of size 0xFF - change the size to 0x100 and have a single global/static buffer in ps1_memory_card.c; sorta already there with `received_game_id`. the card core should receive the game path and transform it into game id in-place, so that then you don't need to call clean_title_id but just grab pointer to that global buffer
- should delete all new `debug_printf` calls
- `mc_pro_flags` should be a volatile int since it's accessed by both cores. Change it to hold a single value rather than bitmask so there are no read-write updates on the simulation core. In `ps1_odeman_task` the behavior should be to retrieve the current value, if it's non-zero => shutdown the simulation core, set to zero (before starting the simulation core again), then do the operation/restart the simulation core.
- in general try to follow the code style of the project specifically opening bracket going to the previous line, spacing around operators (in ps1_odeman.c)

bbsan2k commented:
Updated PR with the following changes:
- Game ID is now still stored twice, but only occupies 0x10 Byte per instance. Cleanup is done before storing it
- Flags are now called command and is volatile and is mutual exclusive - which is ok considering the time it takes to process one command
- debug_printf calls have been removed from time critical func, but remain in ode_man for now
- Changed bracket style to egyptian, operators are still not spaced properly. Is there any clang-format style you are using?

xyzz commented on src/gui.c:
```diff
@@ -703,19 +710,28 @@ void gui_task(void) {
         static char card_idx_s[8];
         static char card_channel_s[8];
-        if (displayed_card_idx != ps1_cardman_get_idx() || displayed_card_channel != ps1_cardman_get_channel()) {
+        // TODO Fix Switching based on game id --> not yet implemented 
```
What does this comment mean?

xyzz commented on src/main.c:
```diff
@@ -55,6 +57,7 @@ int main() {
     check_bootloader_reset();
 
     stdio_uart_init_full(UART_PERIPH, UART_BAUD, UART_TX, UART_RX);
+        //stdio_uart_init_full(UART_PERIPH, UART_BAUD, UART_TX, UART_RX);
```
Leftover? delete this

xyzz commented on src/ps1/ps1_cardman.c:
```diff
@@ -30,6 +32,7 @@ void ps1_cardman_init(void) {
     card_chan = settings_get_ps1_channel();
     if (card_chan < CHAN_MIN || card_chan > CHAN_MAX)
         card_chan = CHAN_MIN;
+    memset(card_game_id, 0, 0x10);
```
this isn't needed, but if you want to keep it anyway use sizeof(card_game_id) instead of 0x10

bbsan2k replied:
Changed it to sizeof. Init should always initialize all variables again - so its kept there

xyzz commented on src/ps1/ps1_cardman.c:
```diff
@@ -73,14 +80,17 @@ void ps1_cardman_open(void) {
 
     ensuredirs();
-    snprintf(path, sizeof(path), "MemoryCards/PS1/Card%d/Card%d-%d.mcd", card_idx, card_idx, card_chan);
+    if ((card_game_id[0] != 0x00) && (card_idx < IDX_MIN)) {
```
the condition is different from the one in `ensuredirs` where it's just checking `card_game_id[0] != 0x00` could that be a problem if there's a mismatch?

bbsan2k replied:
It's better to have the same check in both positions for consistency.

xyzz commented on src/ps1/ps1_cardman.c:
```diff
@@ -164,3 +177,13 @@ int ps1_cardman_get_idx(void) {
 }
+
+void ps1_cardman_set_gameid(const char* game_id) {
+    memcpy(card_game_id, game_id, 0x10);
```
use strlcpy() instead with the size argument set to sizeof(card_game_id)

xyzz commented on src/ps1/ps1_cardman.c:
```diff
@@ -164,3 +177,13 @@ int ps1_cardman_get_idx(void) {
+}
+
+void ps1_cardman_get_gametext(char* const game_text) {
+    memcpy(game_text, card_game_id, 0x10);
```
this function should just return `card_game_id` as a pointer, not do a copy
if it has to do a copy, then it must also take size of the passed `game_text` buffer and use a safe string copy function (strlcpy)

xyzz commented on src/ps1/ps1_memory_card.c:
```diff
@@ -17,13 +20,41 @@ static volatile int reset;
 
 static pio_t cmd_reader, dat_writer;
 static volatile int mc_exit_request, mc_exit_response, mc_enter_request, mc_enter_response;
+static void __time_critical_func(clean_title_id)(const uint8_t* const in_title_id, char* const out_title_id, const size_t in_title_id_length);
```
no need to forward declare this function since it's not being used prior to being defined

xyzz commented on src/ps1/ps1_memory_card.c:
```diff
@@ -17,13 +20,41 @@ static volatile int reset;
+static void __time_critical_func(clean_title_id)(const uint8_t* const in_title_id, char* const out_title_id, const size_t in_title_id_length);
+
+
+static void __time_critical_func(clean_title_id)(const uint8_t* const in_title_id, char* const out_title_id, const size_t in_title_id_length) {
```
this should take size of the output buffer as an argument to make sure there's no overflow

for example:

```
static void __time_critical_func(clean_title_id)(const char *in, size_t in_len, char *out, size_t out_len) {
    size_t j = 0;
    for (size_t i = 0; i < in_len && in[i]; ++i) {
        if (in[i] == '/' || in[i] == '\\') {
            j = 0;
        } else if (j + 1 < out_len) {
            if (in[i] == '_')
                out[j++] = '-';
            else if (in[i] != '.' && in[i] != ';')
                out[j++] = in[i];
        }
    }
    out[j] = '\0';
}
```

bbsan2k replied:
Added parameter. Check was already conducted in while loop condition

xyzz commented on src/ps1/ps1_memory_card.c:
```diff
@@ -148,12 +179,66 @@ static int __time_critical_func(mc_do_state)(uint8_t ch) {
+            switch (byte_count) {
+                case 2: memset(received_game_id, 0, 0x10); return 0x00;
+                case 3: return 0x00;
+                case 4: game_id_length = payload[byte_count - 1]; debug_printf("Length %u\n", game_id_length); return 0x00;
```
remove debug_printf() here

xyzz commented on src/ps1/ps1_memory_card.c:
```diff
@@ -284,4 +369,19 @@ void ps1_memory_card_enter(void) {
     mc_enter_request = mc_enter_response = 0;
     memcard_running = 1;
+    mc_pro_command = 0;
+    game_id_length = 0x10;
+    memset(received_game_id, 0, 0x10);
```
use sizeof(received_game_id) instead of hardcoding 0x10
not sure about the purpose of setting game_id_length initially though?

bbsan2k replied:
Setting it initially is kind of a hack to make sure, that the final condition of reading the payload is not fulfilled to early. Need to find a better solution for this probably

xyzz commented on src/ps1/ps1_odeman.c:
```diff
@@ -0,0 +1,62 @@
+                debug_printf("Received Game ID: ");
+                const char *game_id;
+                game_id = ps1_memory_card_get_game_id();
+                debug_printf("%s\n", game_id);
```
change into a single debug_printf("Received Game ID: %s\n")
